### PR TITLE
mir.sparse

### PR DESCRIPTION
**Pull request by 9il**
PR for tracking #43 implementation.

### CODE REVIEW

**Comment by 9il** [comment]
@wilzbach will use common hash map :-) initial implementation pushed to this branch

File: source/mir/ndslice/sparse.d [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    return .sparse!(T, Lengths.length)([lengths]);
+}
+
+/// ditto
+Sparse!(N, T) sparse(T, size_t N)(auto ref size_t[N] lengths)
+{
+    import mir.ndslice.slice: sliced;
+    T[size_t] table;
+    table[0] = 0;
+    table.remove(0);
```
Comment (wilzbach) [inline source/mir/ndslice/sparse.d]: why is that needed?

Comment (9il) [inline source/mir/ndslice/sparse.d]: to initialize table with not null

Comment (wilzbach) [inline source/mir/ndslice/sparse.d]: but isn't that just a language feature? Or why do you need not null?

```
int[int] a;
writeln(a.get(0, 2)); // 2
```

Comment (9il) [inline source/mir/ndslice/sparse.d]: > writeln(a.get(0, 2)); // 2

writeln(a) ?

Comment (9il) [inline source/mir/ndslice/sparse.d]: assert(a !is null); ?

Comment (9il) [inline source/mir/ndslice/sparse.d]: > but isn't that just a language feature? Or why do you need not null?

For correct behavior when when coping

Comment (wilzbach) [inline source/mir/ndslice/sparse.d]: do you mind to add an unittest for the correct behavior? :)

Comment (9il) [inline source/mir/ndslice/sparse.d]: yes

File: source/mir/ndslice/sparse.d [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+				table.remove(index);
+			}
+		}
+		return value;
+ 	}
+
+	T opIndexOpAssign(string op)(T value, size_t index)
+		if (op == `+` || op == `-`)
+	{
+		mixin (`value = table[index] ` ~ op ~ `= value;`); // this works
```
Comment (wilzbach) [inline source/mir/ndslice/sparse.d]: why not `auto val = table[index]` ~ op ~ `val;`?
In any case, please remove the comment "this works" ;-)

Comment (9il) [inline source/mir/ndslice/sparse.d]: > why not auto val = table[index] ~ op ~ val;?

because this is `?=` ops

> In any case, please remove the comment "this works" ;-)

This is very buggy part for future templated hashmaps

Comment (wilzbach) [inline source/mir/ndslice/sparse.d]: > This is very buggy part for future templated hashmaps

Then state so, at least you draw my immediate attention when you claim "this works".

File: source/mir/ndslice/sparse.d [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+///
+pure unittest
+{
+    auto slice = sparse!double(2, 3);
+    slice[0][] = 1;
+    slice[0, 1] = 2;
+    --slice[0, 0];
+    slice[1, 2] += 4;
+
+    assert(slice == [[0, 2, 1], [0, 0, 4]]);
```
Comment (wilzbach) [inline source/mir/ndslice/sparse.d]: that's works is pretty clever!

File: source/mir/ndslice/sparse.d [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+SUBMODULE = $(LINK2 mir_ndslice_$1.html, mir.ndslice.$1)
+SUBREF = $(LINK2 mir_ndslice_$1.html#.$2, $(TT $2))$(NBSP)
+T2=$(TR $(TDNW $(LREF $1)) $(TD $+))
+T4=$(TR $(TDNW $(LREF $1)) $(TD $2) $(TD $3) $(TD $4))
+
++/
+module mir.ndslice.sparse;
+
+import std.traits;
+import std.meta;
```
Comment (wilzbach) [inline source/mir/ndslice/sparse.d]: you always complained about such global imports and advised me to use selective imports.

Comment (9il) [inline source/mir/ndslice/sparse.d]: traits and meta, somtimes primitives are ok to be public (because they always used)

Comment (9il) [inline source/mir/ndslice/sparse.d]: ... and because I have cleaned them to do not import other modules

**Comment by wilzbach** [comment]
> @wilzbach will use common hash map :-) initial implementation pushed to this branch

I really like the direction this is going and the seamless integration with `Slice`. I would add a couple of unittests that this more, e.g.

```
mySparse[] = mySlice;
mySlice[] = mySparse;
```

and some of the iteration and selection algorithms

**Comment by 9il** [comment]
no, please, let me finish it first. It is faster to implement this than to review api for a couple of small functions

**Comment by wilzbach** [comment]
> no, please, let me finish it first

toggle annoying_bot off

**Comment by 9il** [comment]
> no, please, let me finish it first
> toggle annoying_bot off

Have you start to collect article information for random distributions?

**Comment by 9il** [comment]
It is almost covered and passes CI tests. I have added `sparse_blas` package and moved sum to the mir root.

File: source/mir/ndslice/sparse.d [hunk 5]
```diff
@@ -0,0 +1,623 @@
...
+	T value;
+
+	///
+	sizediff_t opCmp()(auto ref const typeof(this) rht) const
+	{
+		return cmpCoo(this.index, rht.index);
+	}
+}
+
+private sizediff_t cmpCoo(size_t N)(const auto ref size_t[N] a, const auto ref size_t[N] b)
```
Comment (wilzbach) [inline source/mir/ndslice/sparse.d]: you only use this once (three lines above) - thus why not directly putting it there?

Comment (9il) [inline source/mir/ndslice/sparse.d]: this may be used by the future algorithms

File: source/mir/ndslice/sparse.d [hunk 6]
```diff
@@ -0,0 +1,623 @@
...
+				index %= _strides[i];
+			}
+			ret.index[N - 1] = index;
+			ret.value = iv.value;
+			return ret;
+		}
+
+		//int opApply(Dg)(scope Dg dg)
+		//	if (ParameterTypeTuple!Dg.length == N + 1 && allSatisfy!(isIndex, ParameterTypeTuple!Dg[0..N]))
+		//{
```
Comment (wilzbach) [inline source/mir/ndslice/sparse.d]: I guess you remove that before merging? Or do you still work on it?

File: source/mir/ndslice/sparse.d [hunk 7]
```diff
@@ -0,0 +1,623 @@
...
+		return _length;
+	}
+}
+
+///
+auto	compress
+	(I = uint, J = uint, S : Slice!(N, R), size_t N, R)
+	(S slice)
+	if(N > 1)
+{
```
Comment (wilzbach) [inline source/mir/ndslice/sparse.d]: weird formatting

File: source/mir/ndslice/sparse.d (hunk 6 above)
Comment (9il) [inline source/mir/ndslice/sparse.d]: will remove

**Comment by wilzbach** [comment]
> It is almost covered and passes CI tests. I have added sparse_blas package and moved sum to the mir root.

Wow! Looks great on my first pass. I will have to play with to give further feedback (will do so tonight).

Nitpicks from the first pass:
- a lot of doc is missing
- your editor is using tabs? Please configure it to use spaces ;-)
- saw a couple of places with debug code

Do you plan to support the `nogc` code via Allocators?
This probably requires a custom implementation of hashmaps?

> moved sum to the mir root.

you probably have to update the `index.d` too

**Comment by wilzbach** [comment]
btw if you want to upload docs for a branch to s3 - that's pretty easy:

```
export AWS_ACCESS_KEY_ID=XYZ; export AWS_SECRET_ACCESS_KEY=XYZ
rdmd bootDoc/generate.d --output=docs source --separator=_ --extra index.d
AWS_DEFAULT_REGION=eu-west-1 aws s3 sync --acl public-read --delete docs s3://docs.mir.dlang.io/sparse
```

**Comment by 9il** [comment]
> Do you plan to support the nogc code via Allocators?

Yes, but when it would be at phobos.

> This probably requires a custom implementation of hashmaps?

Yes for `Sparse` and no for `CompressedTensor`.

**Comment by 9il** [comment]
Please forgot this keys OO

**Comment by 9il** [comment]
Where you found AWS keys?

**Comment by 9il** [comment]
i moved sparse and blas to separate package, so we would not include it to ndslice. I just added few ptr methods to ndslice.

File: source/mir/ndslice/slice.d [hunk 8]
```diff
@@ -1150,30 +1150,60 @@ struct Slice(size_t _N, _Range)
...
+        _ptr = shell;
     }
 
-    static if (!hasPtrBehavior!PureRange)
-    this(ref in size_t[PureN] lengths, ref in sizediff_t[PureN] strides, PtrShell!PureRange shell)
+    public:
+
+    /++
+    This constructor should be used only for integration with other languages or libraries such as Julia and numpy.
+    Params:
```
Comment (wilzbach) [inline source/mir/ndslice/slice.d:1167]: thanks for changing & documenting this :)

File: source/mir/sparse/blas/axpy.d [hunk 9]
```diff
@@ -0,0 +1,70 @@
+module mir.sparse.blas.axpy;
+
+import std.traits;
+import mir.ndslice.slice;
+import mir.sparse;
+
+/++
```
Comment (wilzbach) [inline source/mir/sparse/blas/axpy.d:12]: description?

File: source/mir/sparse/blas/dot.d [hunk 10]
```diff
@@ -0,0 +1,173 @@
+module mir.sparse.blas.dot;
+
+import std.traits;
+import mir.ndslice.slice;
+import mir.sparse;
+
+
+/++
+Params:
```
Comment (wilzbach) [inline source/mir/sparse/blas/dot.d:14]: description ;-)

File: source/mir/sparse/blas/dot.d [hunk 11]
```diff
@@ -0,0 +1,173 @@
...
+(V1 x, V2 y)
+	if (isDynamicArray!V2 || is(V2 : Slice!(1, V2R), V2R))
+in
+{
+	if(x.indexes.length)
+		assert(x.indexes[$-1] < y.length);
+}
+body
+{
+	pragma(inline, false);
```
Comment (wilzbach) [inline source/mir/sparse/blas/dot.d]: here are still tabs :/

File: source/mir/sparse/blas/gemm.d [hunk 12]
```diff
@@ -0,0 +1,141 @@
+module mir.sparse.blas.gemm;
+
+import std.traits;
+import mir.ndslice.slice;
+import mir.sparse;
+
+/++
```
Comment (wilzbach) [inline source/mir/sparse/blas/gemm.d:12]: description?

File: source/mir/sparse/package.d [hunk 13]
```diff
@@ -0,0 +1,614 @@
+/++
+$(H2 Sparse Tensors)
+
+This is a submodule of $(LINK2 mir_ndslice.html, mir.ndslice).
+
+
```
Comment (wilzbach) [inline source/mir/sparse/package.d:6]: two empty lines

File: source/mir/sparse/package.d [hunk 14]
```diff
@@ -0,0 +1,614 @@
...
+import std.traits;
+import std.meta;
+
+import mir.ndslice.slice;
+
+private enum isIndex(I) = is(I : size_t);
+
+//TODO: replace with `static foreach`
+private template Iota(size_t i, size_t j)
```
Comment (wilzbach) [inline source/mir/sparse/package.d:27]: couldn't that be useful for `Iota` from `ndslice`, too?

File: source/mir/sparse/package.d [hunk 15]
```diff
@@ -0,0 +1,614 @@
...
+                table.remove(index);
+            }
+        }
+        return value;
+    }
+}
+
+/++
++/
+struct CoordinateValue(size_t N, T)
```
Comment (wilzbach) [inline source/mir/sparse/package.d:176]: document me

File: source/mir/sparse/package.d [hunk 16]
```diff
@@ -0,0 +1,614 @@
...
+{
+    foreach(i; Iota!(0, N))
+        if(auto d = a[i] - b[i])
+            return d;
+    return 0;
+}
+
+/++
++/
+auto byCoordinateValue(S : Slice!(N, R), size_t N, R : SparseMap!T, T)(S slice)
```
Comment (wilzbach) [inline source/mir/sparse/package.d:204]: document me

File: source/mir/sparse/package.d [hunk 17]
```diff
@@ -0,0 +1,614 @@
+/++
+$(H2 Sparse Tensors)
+
+This is a submodule of $(LINK2 mir_ndslice.html, mir.ndslice).
```
Comment (wilzbach) [inline source/mir/sparse/package.d:4]: not anymore?

File: source/mir/sparse/package.d [hunk 18]
```diff
@@ -0,0 +1,614 @@
...
+}
+
+/++
+Sparse tensors represented in Dictionary of Keys (DOK) format.
+
+Params:
+    N = dimension count
+    lengths = list of dimension lengths
+Returns:
+    `N`-dimensional slice composed of indexes
```
Comment (wilzbach) [inline source/mir/sparse/package.d]: nitpick: no need to escape `N` and add a space ;-)

File: source/mir/sparse/package.d [hunk 19]
```diff
@@ -0,0 +1,614 @@
...
+auto byCoordinateValue(S : Slice!(N, R), size_t N, R : SparseMap!T, T)(S slice)
+{
+    static struct CoordinateValues
+    {
+        static if(N > 1)
+            private sizediff_t[N-1] _strides;
+        mixin _sparse_range_methods!(N, T);
+        private typeof(Sparse!(N, T).init.ptr.range.table.byKeyValue()) _range;
+
+        auto front() @property
```
Comment (wilzbach) [inline source/mir/sparse/package.d]: empty `@property` properties should be `const`

File: source/mir/sparse/package.d [hunk 20]
```diff
@@ -0,0 +1,614 @@
...
+        alias Iota = AliasSeq!(i, Iota!(i + 1, j));
+}
+
+/++
+Sparse tensors represented in Dictionary of Keys (DOK) format.
+
+Params:
+    N = dimension count
+    lengths = list of dimension lengths
+Returns:
```
Comment (wilzbach) [inline source/mir/sparse/package.d:42]: nitpick: empty line between Params and Returns

File: source/mir/sparse/package.d [hunk 21]
```diff
@@ -0,0 +1,614 @@
...
+}
+
+/// ditto
+Sparse!(N, T) sparse(T, size_t N)(auto ref size_t[N] lengths)
+{
+    import mir.ndslice.slice: sliced;
+    T[size_t] table;
+    table[0] = 0;
+    table.remove(0);
+    assert(table !is null);
```
Comment (wilzbach) [inline source/mir/sparse/package.d]: that's is still so weird - it really does make a difference (I just checked myself). Could you please leave a comment there - probably the people at phobos will complain otherwise too ;-)

File: source/mir/sparse/package.d [hunk 22]
```diff
@@ -0,0 +1,614 @@
...
+    }
+
+    ///
+    T opIndex(size_t index)
+    {
+        static if (isScalarType!T)
+            return table.get(index, cast(T)0);
+        else
+            return table.get(index, null);
+    }
```
Comment (wilzbach) [inline source/mir/sparse/package.d]: do you want to support `opCall` again for the Math & Fortran order?

File: source/mir/sparse/package.d [hunk 23]
```diff
@@ -0,0 +1,614 @@
...
+            }
+        }
+        return value;
+    }
+
+    ///
+    T opIndexOpAssign(string op)(T value, size_t index)
+        if (op == `+` || op == `-`)
+    {
+        mixin (`value = table[index] ` ~ op ~ `= value;`); // this works
```
Comment (wilzbach) [inline source/mir/sparse/package.d]: as previously mentioned - please remove `this works` ;-)

File: source/mir/sparse/package.d [hunk 24]
```diff
@@ -0,0 +1,614 @@
...
+    auto slice = sparse!double(3, 3);
+    slice[] = [[0, 2, 1], [0, 0, 4], [6, 7, 0]];
+    assert(slice.byValueOnly.array.sort().release == [1, 2, 4, 6, 7]);
+}
+
+private mixin template _sparse_range_methods(size_t N, T)
+{
+    private size_t _length;
+
+    void popFront()
```
Comment (wilzbach) [inline source/mir/sparse/package.d]: is `@safe`, `nothrow` and `pure` automatically added here?

File: source/mir/sparse/package.d [hunk 25]
```diff
@@ -0,0 +1,614 @@
...
+        _length--;
+    }
+
+    bool empty() const @property
+    {
+        return _length == 0;
+    }
+
+    auto save() @property
+    {
```
Comment (wilzbach) [inline source/mir/sparse/package.d]: zero const `@property` should be `const` - or not a `@property` ;-)

File: source/mir/sparse/package.d [hunk 26]
```diff
@@ -0,0 +1,614 @@
...
+///
+pure unittest
+{
+    import std.array: array;
+    import std.algorithm.sorting: sort;
+    alias CV = CoordinateValue!(2, double);
+
+    auto slice = sparse!double(3, 3);
+    slice[] = [[0, 2, 1], [0, 0, 4], [6, 7, 0]];
+    assert(slice.byCoordinateValue.array.sort().release == [
```
Comment (wilzbach) [inline source/mir/sparse/package.d]: maybe we should have a `flag` for `Sorted.yes` as sorting could be implemented more efficient?

File: source/mir/sparse/package.d [hunk 27]
```diff
@@ -0,0 +1,614 @@
...
+    else
+        return Coordinates(slice.ptr.range.table.byKey);
+}
+
+///
+pure unittest
+{
+    import std.array: array;
+    import std.algorithm.sorting: sort;
+    alias CV = CoordinateValue!(2, double);
```
Comment (wilzbach) [inline source/mir/sparse/package.d]: copy paste error

File: source/mir/sparse/package.d [hunk 28]
```diff
@@ -0,0 +1,614 @@
...
+{
+    return compressWithType!(DeepElementType!(Slice!(N, R)))(slice);
+}
+
+/// Sparse tensor compression
+unittest
+{
+    import std.array: array;
+    import std.algorithm.sorting: sort;
+    alias CV = CoordinateValue!(2, double);
```
Comment (wilzbach) [inline source/mir/sparse/package.d]: copy paste error?

File: source/mir/sparse/package.d [hunk 29]
```diff
@@ -0,0 +1,614 @@
...
+        [1, 2, 2, 0, 2, 2],
+        [0, 2, 3, 3, 5, 6]));
+}
+
+/// Sparse tensor compression
+unittest
+{
+    import std.array: array;
+    import std.algorithm.sorting: sort;
+    alias CV = CoordinateValue!(2, double);
```
Comment (wilzbach) [inline source/mir/sparse/package.d]: copy paste error?

File: source/mir/sparse/package.d [hunk 30]
```diff
@@ -0,0 +1,614 @@
...
+    }
+
+    size_t length() const @property
+    {
+        return _length;
+    }
+}
+
+///
+auto compress(I = uint, J = uint, S : Slice!(N, R), size_t N, R)(S slice)
```
Comment (wilzbach) [inline source/mir/sparse/package.d:378]: needs docs ;-)

File: source/mir/sparse/package.d [hunk 31]
```diff
@@ -0,0 +1,614 @@
...
+        [1, 7, 7, 0, 7, 7],
+        [0, 2, 3, 3, 5, 6]));
+}
+
+/// Dense tensor compression
+unittest
+{
+    import std.array: array;
+    import std.algorithm.sorting: sort;
+    alias CV = CoordinateValue!(2, double);
```
Comment (wilzbach) [inline source/mir/sparse/package.d]: ditto

File: source/mir/sparse/package.d [hunk 32]
```diff
@@ -0,0 +1,614 @@
...
+        [1, 2, 2, 0, 2, 2],
+        [0, 2, 3, 3, 5, 6]));
+}
+
+/// Dense tensor compression
+unittest
+{
+    import std.array: array;
+    import std.algorithm.sorting: sort;
+    alias CV = CoordinateValue!(2, double);
```
Comment (wilzbach) [inline source/mir/sparse/package.d]: ditto

File: source/mir/sparse/package.d [hunk 33]
```diff
@@ -0,0 +1,614 @@
...
+    import std.algorithm.sorting: sort;
+    import mir.ndslice.selection: iotaSlice;
+    auto data = slice
+        .ptr
+        .range
+        .table
+        .byKeyValue
+        .array
+        .sort!((a, b) => a.key < b.key)
+        .release;
```
Comment (wilzbach) [inline source/mir/sparse/package.d]: this calls [move](https://dlang.org/phobos/std_algorithm_mutation.html#.move) under the hood - doesn't this yield an unnecessary copy?

File: source/mir/sparse/package.d [hunk 34]
```diff
@@ -0,0 +1,614 @@
...
+            map.pointers[k + 2 .. p + 2] = map.pointers[k + 1];
+            k = p;
+        }
+        map.pointers[k + 1]++;
+    }
+    map.pointers[k + 2 .. $] = map.pointers[k + 1];
+    return map.sliced(inv.shape);
+}
+
+
```
Comment (wilzbach) [inline source/mir/sparse/package.d:509]: two empty lines

File: source/mir/sparse/package.d [hunk 35]
```diff
@@ -0,0 +1,614 @@
...
+    import std.array: array;
+    import std.algorithm.sorting: sort;
+    import mir.ndslice.selection: iotaSlice;
+    auto data = slice
+        .ptr
+        .range
+        .table
+        .byKeyValue
+        .array
+        .sort!((a, b) => a.key < b.key)
```
Comment (wilzbach) [inline source/mir/sparse/package.d]: see remark about using a flag for sorted output

File: source/mir/sparse/package.d [hunk 36]
```diff
@@ -0,0 +1,614 @@
...
+{
+    import std.array: array;
+    import std.algorithm.sorting: sort;
+    import mir.ndslice.selection: iotaSlice;
+    auto data = slice
+        .ptr
+        .range
+        .table
+        .byKeyValue
+        .array
```
Comment (wilzbach) [inline source/mir/sparse/package.d]: can't the byKeyValue `struct` proxy the index operations? then you don't an allocation here.

File: source/mir/sparse/package.d (hunk 30 above)
Comment (wilzbach) [inline source/mir/sparse/package.d:378]: https://en.wikipedia.org/wiki/Sparse_matrix#Compressed_sparse_row_.28CSR.2C_CRS_or_Yale_format.29

File: source/mir/sparse/package.d [hunk 37]
```diff
@@ -0,0 +1,614 @@
...
+        .sort!((a, b) => a.key < b.key)
+        .release;
+    auto inv = iotaSlice(slice.shape[0 .. N - 1]);
+    auto count = inv.elementsCount;
```
Comment (wilzbach) [inline source/mir/sparse/package.d]: is there no better way to count the number of rows? should be possible directly with `shape`?

File: source/mir/sparse/package.d [hunk 38]
```diff
@@ -0,0 +1,614 @@
...
+        );
+    size_t k = 0;
+    map.pointers[0] = 0;
+    map.pointers[1] = 0;
+    foreach(t, e; data)
+    {
+        map.values[t] = cast(V) e.value;
+        size_t index = e.key;
+        map.indexes[t] = cast(I)(index % slice.length!(N - 1));
+        auto p = index / slice.length!(N - 1);
```
Comment (wilzbach) [inline source/mir/sparse/package.d]: maybe we can store the result of `slice.length!(N - 1) in one variable? would cleanup the loop.

File: source/mir/sparse/package.d [hunk 39]
```diff
@@ -0,0 +1,614 @@
...
+        if(k != p)
```
Comment (wilzbach) [inline source/mir/sparse/package.d]: maybe add doc: "// is in a row?"

File: source/mir/sparse/blas/axpy.d (hunk 9 above)
Comment (wilzbach) [inline source/mir/sparse/blas/axpy.d:12]: "constant times a vector plus a vector"

File: source/mir/sparse/blas/axpy.d [hunk 40]
```diff
@@ -0,0 +1,70 @@
...
+	}
+}
+
+///
+unittest
+{
+	auto x = CompressedArray!(double, uint)([1.0, 3, 4, 9, 13], [0, 3, 5, 9, 10]);
+	auto y = [0.0, 1.0, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12];
+	axpy(2.0, x, y);
+	assert(y == [2.0, 1.0, 2, 9, 4, 13, 6, 7, 8, 27, 36, 11, 12]);
```
Comment (wilzbach) [inline source/mir/sparse/blas/axpy.d]: second value: 3* 2.0 + 1 = 7?

File: source/mir/sparse/blas/dot.d (hunk 10 above)
Comment (wilzbach) [inline source/mir/sparse/blas/dot.d:14]: forms the dot product of two vectors

Comment (wilzbach) [inline source/mir/sparse/blas/dot.d:14]: https://en.wikipedia.org/wiki/Dot_product

File: source/mir/sparse/blas/dot.d [hunk 41]
```diff
@@ -0,0 +1,173 @@
...
+
+	return s;
+}
+
+///
+unittest
+{
+	auto x = CompressedArray!(int, uint)([1, 3, 4, 9, 10], [0, 3, 5, 9, 100]);
+	auto y = CompressedArray!(int, uint)([1, 10, 100, 1000], [1, 3, 4, 9]);
+	assert(dot(x, y) == 9030);
```
Comment (wilzbach) [inline source/mir/sparse/blas/dot.d:111]: maybe you want to add the vector representation as comment - it's a lot easier to read ;-)

```
x = [1,0,0, 3,0,4,0,0,0,9,.....,10]
y = [0,1,0,10,0,0,0,0,0,1000]
```

= 30 + 9000

Alternatively use `compress` ;-)

File: source/mir/sparse/blas/dot.d [hunk 42]
```diff
@@ -0,0 +1,173 @@
...
+	}
+
+	return s;
+}
+
+///
+unittest
+{
+	import std.typecons: No;
+	auto x = CompressedArray!(double, uint)([1.0, 3, 4, 9, 13], [0, 3, 5, 9, 10]);
```
Comment (wilzbach) [inline source/mir/sparse/blas/dot.d:173]: easier to read (maybe as comment?)

```
x = [1,0,0,3,0,4,0,0,0,9,13,  0,  0,  0]
y = [0,1,2,3,4,5,6,7,8,9,10,11,12, 13]
```

File: source/mir/sparse/blas/dot.d [hunk 43]
```diff
@@ -0,0 +1,173 @@
...
+	auto y = [0.0, 1.0, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12];
+	auto r = 0 + 3 * 3 + 5 * 4 + 9 * 9 + 10 * 13;
```
Comment (wilzbach) [inline source/mir/sparse/blas/dot.d:175]: why do you multiply `y` first? .e.g `5*4` or `10 * 13`?

**Comment by wilzbach** [comment]
1) byCoordinate, byCoordinateValue, byValueOnly share too much common code.

Maybe a common struct with `static if` is better than mixins here? Even the `front` method are very similar.

2) I still saw a lot of tabs
3) I really like that you created a file for every blas method!
4) do you plan to add the blas methods a overloads to sliced and sparse once they are stable? e.g. a \* b could use `dot`?

File: source/mir/sparse/package.d (hunk 14 above)
Comment (9il) [inline source/mir/sparse/package.d:27]: Just small copy past. This is ok

File: source/mir/sparse/package.d (hunk 17 above)
Comment (9il) [inline source/mir/sparse/package.d:4]: yes, should be fixed

File: source/mir/sparse/package.d (hunk 19 above)
Comment (9il) [inline source/mir/sparse/package.d]: If this returns a class than class would be const and user would not be able to change it

File: source/mir/sparse/package.d (hunk 22 above)
Comment (9il) [inline source/mir/sparse/package.d]: This os works out of the box using `Sparse`

File: source/mir/sparse/package.d (hunk 23 above)
Comment (9il) [inline source/mir/sparse/package.d]: i need to show attention to this line

File: source/mir/sparse/package.d (hunk 24 above)
Comment (9il) [inline source/mir/sparse/package.d]: yes

File: source/mir/sparse/package.d (hunk 25 above)
Comment (9il) [inline source/mir/sparse/package.d]: Again, this would not work with reference types

File: source/mir/sparse/package.d (hunk 26 above)
Comment (9il) [inline source/mir/sparse/package.d]: i don't understand

File: source/mir/sparse/package.d (hunk 22 above)
Comment (wilzbach) [inline source/mir/sparse/package.d]: Oh right - awesome :+1:

File: source/mir/sparse/package.d (hunk 33 above)
Comment (9il) [inline source/mir/sparse/package.d]: no

File: source/mir/sparse/package.d (hunk 26 above)
Comment (wilzbach) [inline source/mir/sparse/package.d]: `slice.byCoordinateValue!Sorted.yes`

File: source/mir/sparse/package.d (hunk 36 above)
Comment (9il) [inline source/mir/sparse/package.d]: no, it is not sorted

File: source/mir/sparse/package.d (hunk 37 above)
Comment (9il) [inline source/mir/sparse/package.d]: `elementsCount` works directly with shape

**Comment by 9il** [comment]
> 4) do you plan to add the blas methods a overloads to sliced and sparse once they are stable? e.g. a \* b could use dot?

No, this is buggy approach for D

**Comment by 9il** [comment]
@wilzbach I have added  dense BLAS naive implementation small `mir.math` and `mir.sparse.blas.gemm: selectiveGemm`

**Comment by 9il** [comment]
BTW, dense blas is very slow))

File: source/mir/blas/dot.d [hunk 44]
```diff
@@ -0,0 +1,127 @@
...
+Note:
+	`dot` implementation is naive for now.
++/
+auto dot(V1, V2)(scope V1 x, scope V2 y)
+	if ((isDynamicArray!V1 || is(V1 : Slice!(1, T1*), T1)) &&
+		(isDynamicArray!V2 || is(V2 : Slice!(1, T2*), T2)) &&
+		is(Unqual!(ForeachType!V1) == Unqual!(ForeachType!V2)))
+in
+{
+	assert(x.length == y.length);
```
Comment (wilzbach) [inline source/mir/blas/dot.d:26]: what is our consensus on throwing exceptions on user input?

File: source/mir/blas/dot.d [hunk 45]
```diff
@@ -0,0 +1,127 @@
...
+	}
+	else
+	static if(!isDynamicArray!V1 && isDynamicArray!V2)
+	{
+		import mir.internal.utility;
+		return
+			x.stride == 1 ?
+				dotImpl(x.toDense, y) :
+				dotImpl(y, x);
+	}
```
Comment (wilzbach) [inline source/mir/blas/dot.d:52]: Can't you combine those two cases with alias?

```
alias f1 = V1, f2 = V2;

static if(isDynamicArray!V1 && !isDynamicArray!V2)
{
   f1 = V2, f2= V1;
}
static if(!isDynamicArray!V1 && isDynamicArray!V2)
{
                import mir.internal.utility;
        return
            x.stride == 1 ?
                dotImpl(x.toDense, y) :
                dotImpl(y, x);
}
```

File: source/mir/internal/math.d [hunk 46]
```diff
@@ -0,0 +1,67 @@
...
+{
+	import ldc.intrinsics;
+
+	alias sqrt = llvm_sqrt;
+
+	alias sin = llvm_sin;
+
+	alias cos = llvm_cos;
+
+	//alias powi = llvm_powi;
```
Comment (wilzbach) [inline source/mir/internal/math.d]: ?

File: source/mir/blas/dot.d [hunk 47]
```diff
@@ -0,0 +1,127 @@
...
+	y = vector
+Returns:
+	scalar `xᵀ × y`
+Note:
+	`dot` implementation is naive for now.
++/
+auto dot(V1, V2)(scope V1 x, scope V2 y)
+	if ((isDynamicArray!V1 || is(V1 : Slice!(1, T1*), T1)) &&
+		(isDynamicArray!V2 || is(V2 : Slice!(1, T2*), T2)) &&
+		is(Unqual!(ForeachType!V1) == Unqual!(ForeachType!V2)))
```
Comment (wilzbach) [inline source/mir/blas/dot.d:23]: maybe combining `isDynamicArray!V2 || is(V2 : Slice!(1, T2*), T2)` in a template?
Currently it's not very readable for a user ...

File: source/mir/blas/README.md [hunk 48]
```diff
@@ -0,0 +1 @@
+BLAS is not ready. It is slow. It is not tested. Cache oriented BLAS would be anonced at https://forum.dlang.org .
```
Comment (wilzbach) [inline source/mir/blas/README.md]: "BLAS is not ready" -> "BLAS is still in development and will undergo more work & testing. Use at your own risk. Cache oriented BLAS would be anonced at https://forum.dlang.org and [Twitter](https://twitter.com/libmir).

File: source/mir/internal/math.d (hunk 46 above)
Comment (9il) [inline source/mir/internal/math.d]: Use the same name like in `std.math`

File: source/mir/blas/dot.d (hunk 45 above)
Comment (9il) [inline source/mir/blas/dot.d:52]: BLAS code is draft. This should not be reduced in code size because it would slowdown development process.

File: source/mir/blas/dot.d (hunk 44 above)
Comment (9il) [inline source/mir/blas/dot.d:26]: Only if consistency check is very hard for user like in `reshape`

File: source/mir/blas/dot.d (hunk 47 above)
Comment (wilzbach) [inline source/mir/blas/dot.d:23]: have you tried?

```
enum bool isVector(V, T) = (isDynamicArray!V || is(V : Slice!(1, T*), T));
```

File: source/mir/internal/math.d [hunk 49]
```diff
@@ -0,0 +1,83 @@
...
+
+	alias rint = llvm_rint;
+
+	alias nearbyint = llvm_nearbyint;
+
+	alias copysign = llvm_copysign;
+
+	alias round = llvm_round;
+
+    alias fmuladd = llvm_fmuladd;
```
Comment (wilzbach) [inline source/mir/internal/math.d:47]: nitpick - you combine white space and tabs in here :/

File: source/mir/internal/utility.d [hunk 50]
```diff
@@ -0,0 +1,57 @@
...
+		return ptr.range[ptr.shift .. x.length + ptr.shift];
+	}
+}
+
+alias ConstData(S : Slice!(N, T*), size_t N, T) = Slice!(N, const(T)*);
+alias ConstData(S : Slice!(N, T[]), size_t N, T) = Slice!(N, const(T)[]);
+alias ConstData(S : T[], T) = const(T)[];
+
+enum isVector(V) = is(V : T1[], T1) || is(V : Slice!(1, T2*), T2);
+enum isMatrix(M) = is(M : Slice!(2, T*), T);
```
Comment (wilzbach) [inline source/mir/internal/utility.d:41]: please document these templates

File: source/mir/internal/utility.d [hunk 51]
```diff
@@ -0,0 +1,57 @@
...
+//	is(Slice!(1, const    (T)* )) ||
+//	is(Slice!(1, immutable(T)* )) ;
+
+//enum isConstMatrix(T) ==
+//	is(Slice!(2,           T [])) ||
+//	is(Slice!(2, const    (T)[])) ||
+//	is(Slice!(2, immutable(T)[])) ||
+//	is(Slice!(2,           T * )) ||
+//	is(Slice!(2, const    (T)* )) ||
+//	is(Slice!(2, immutable(T)* )) ;
```
Comment (wilzbach) [inline source/mir/internal/utility.d:57]: I thought having commented code is against our coding policy :-1:
